### DISCUSSION TITLE
Cannot link to proper libraries on Windows 11 Pro.

### DISCUSSION THREAD

**Discussion by huskier** [libvips/libvips#3944] [Q&A] [answered]
We want to use libvips to read and write images in our Qt application on Windows 11 Pro with MSVC2019 toolkit (x64). We've downloaded vips-dev-8.15.

We've dealed the include headers, and the application have passed compilation.

For the linker, we have told the linker to link the vips libraries: 
LIBS += -LD:/development/github_gitlib/vips-dev-8.15/lib -llibvips-cpp -llibvips

However, we still got "unresolved external symbos" errors as the following output.
> "__declspec(dllimport) public: class vips::VOption * __cdecl vips::VOption::set(char const *,int)"
> "__declspec(dllimport) public: static class vips::VOption * __cdecl vips::VImage::option(void)"
> "__declspec(dllimport) public: static class vips::VImage __cdecl vips::VImage::new_from_file
> "__declspec(dllimport) public: __cdecl vips::VImage::~VImage(void)"

Are there any needed libraries we forgot to link?

The code snippet in our source is as following:
```
    VImage in = VImage::new_from_file ("D:/development/DCS/tests/pcb.png", VImage::option ()->set ("access", VIPS_ACCESS_SEQUENTIAL));
    vips_shutdown ();
```

**Comment by jcupitt** [comment, marked as answer]
Hello @huskier,

The C++ binary we provide for windows is for gcc 15. C++ has no ABI, so you have to match the compiler and compiler version exactly along your whole toolchain.

If you are using MSVC, you will need to rebuild the libvips C++ binding yourself. It's very easy -- just copy these files into your project:

https://github.com/libvips/libvips/tree/master/cplusplus

**Reply by huskier** [reply]
Thank you for your reply,  @jcupitt.

My understanding is that we do not need to link the libvips-cpp.lib, if we have all the cpp source code embedded into our project. However, after copying those cpp files into our project, we've gotten more "unresolved external symbos" errors (78 more).

Something I forget?

We've confirmed that the following C API interface could be compiled, linked and run without problem.
>     VipsImage *in;
>     in = vips_image_new_from_file("D:/development/DCS/tests/pcb.png", NULL);

**Reply by huskier** [reply]
We've gotten the solution. 
1. #define VIPS_CPLUSPLUS_EXPORTS
2. Appending "-llibglib-2.0 -llibgobject-2.0" to the following line.
`LIBS += -LD:/development/github_gitlib/vips-dev-8.15/lib -llibvips -llibglib-2.0 -llibgobject-2.0
`
Finally, the errors have gone!

Now, we could happily use VIPS, and we like it very very much!

**Reply by jcupitt** [reply]
I'm glad it's working! Thanks for the update.

**Reply by kleisauke** [reply]
> 1. #define VIPS_CPLUSPLUS_EXPORTS

For further readers, `VIPS_CPLUSPLUS_EXPORTS` is gone since libvips 8.16, see:
https://github.com/libvips/libvips/pull/4033
https://github.com/libvips/libvips/issues/4270#issuecomment-2486094757
